platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,754 @@
+            return Err(sdmmc::Error::Busy);
+        }
+
+        let cmd_idx = params.index;
+        // TODO: naive `auto_stop` selection, this depends on the command (argument) that is used.
+        // Should this (auto_stop select) be part of the params?
+        let (data_trans, trans_dir, auto_stop) = match params.kind {
+            sdmmc::CommandKind::Control => (false, false, false),
+            sdmmc::CommandKind::Read(len) => (true, false, true),
+            sdmmc::CommandKind::Write(len) => (true, true, true),
+        };
+        let chk_resp_crc = params.rsp_crc;
+        let long_resp = params.rsp_type == sdmmc::ResponseType::Long;
+        let resp_rcv = params.rsp_type != sdmmc::ResponseType::None;
+
+        // Do any required configuration
+        match params.options {
+            sdmmc::HardwareOptions::None => (),
+            sdmmc::HardwareOptions::SetBusWidth(width) => {
+                self.smhc.smhc_ctype.write(|w| match width {
+                    sdmmc::BusWidth::Single => w.card_wid().b1(),
+                    sdmmc::BusWidth::Quad => w.card_wid().b4(),
+                    sdmmc::BusWidth::Octo => w.card_wid().b8(),
+                });
+            }
+        }
+
+        let mut guard = self.isr.lock(self.smhc);
+        guard.data.op = match (params.kind, params.buffer) {
+            (sdmmc::CommandKind::Control, _) => SmhcOp::Control,
+            (sdmmc::CommandKind::Read(cnt), Some(buf)) => SmhcOp::Read {
+                buf,
+                cnt,
+                auto_stop,
+            },
+            (sdmmc::CommandKind::Write(cnt), Some(buf)) => SmhcOp::Write {
+                buf,
+                cnt,
+                auto_stop,
+            },
+            _ => {
+                tracing::error!("did not provide a buffer for read/write");
+                return Err(sdmmc::Error::Other);
+            }
+        };
+
+        // Statically allocate space for 16 DMA descriptors.
+        // Each descriptor can do a transfer of 4KB, giving a max total transfer of 64KB.
+        // By declaring them in this scope the descriptor memory will live long enough,
+        // however currently this is up to the user to guarantuee.
+        // Safety: a zero'ed descriptor is valid and will simply be ignored by the IDMAC.
+        let mut descriptors: [idmac::Descriptor; 16] = unsafe { core::mem::zeroed() };
+
+        // Perform checks on arguments to make sure we won't overflow the buffer
+        match &guard.data.op {
+            SmhcOp::Read { buf, cnt, .. } | SmhcOp::Write { buf, cnt, .. } => {
+                const DESCR_BUFF_SIZE: usize = 0x1000;
+
+                // Currently we limit the number of data that can be read at once
+                if *cnt > DESCR_BUFF_SIZE * descriptors.len() || buf.capacity() < *cnt {
+                    return Err(sdmmc::Error::Data);
+                }
+
+                tracing::debug!("Creating descriptor chain from buffer");
```

Comment:
nit, take it or leave it: maybe worth including the descriptor count?
```suggestion
                tracing::debug!(cnt, "Creating descriptor chain from buffer");
```